src/Stratis.Bitcoin/P2P/PeerSelector.cs

```diff
@@ -16,55 +16,144 @@ public interface IPeerSelector
+        IEnumerable<PeerAddress> SelectPeers(int peerCount);
     }
 
     public sealed class PeerSelector : IPeerSelector
     {
+        /// <summary>Logger factory to create loggers.</summary>
+        private readonly ILoggerFactory loggerFactory;
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
         /// <summary>
         /// The address manager instance that holds the peer list to be queried. 
         /// </summary>
         private readonly ConcurrentDictionary<IPEndPoint, PeerAddress> peerAddresses;
 
+        /// <summary>Random number generator used when selecting and ordering peers.</summary>
+        private readonly Random random;
+
         /// <summary>
         /// Constructor for the peer selector.
         /// </summary>
         /// <param name="peerAddresses">The collection of peer address as managed by the peer address manager.</param>
-        public PeerSelector(ConcurrentDictionary<IPEndPoint, PeerAddress> peerAddresses)
+        public PeerSelector(ILoggerFactory loggerFactory, ConcurrentDictionary<IPEndPoint, PeerAddress> peerAddresses)
         {
-            Guard.NotNull(peerAddresses, nameof(peerAddresses));
-
+            this.loggerFactory = loggerFactory;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
             this.peerAddresses = peerAddresses;
+            this.random = new Random();
         }
 
         /// <inheritdoc/>
         public PeerAddress SelectPeer()
         {
-            var tried = this.peerAddresses.Attempted().Concat(this.peerAddresses.Connected());
-            if (tried.Any() == true && (!this.peerAddresses.Fresh().Any() || new Random().Next(2) == 0))
-                return tried.Random();
+            this.logger.LogTrace("()");
+
+            PeerAddress peerAddress = null;
+
+            var peers = this.SelectPreferredPeers();
+            if (peers.Any())
+            {
+                peerAddress = peers.Random();
+                this.logger.LogTrace("(-):'{0}'", peerAddress.NetworkAddress.Endpoint);
+            }
+            else
+                this.logger.LogTrace("(-)[NO_PEER]");
+
+            return peerAddress;
+        }
+
+        /// <summary>
+        /// Filtering logic for selecting peers to connect to via the <see cref="PeerConnector"/> classes.
+        /// </summary>
+        private IEnumerable<PeerAddress> SelectPreferredPeers()
+        {
+            this.logger.LogTrace("()");
+
+            // First check to see if there are handshaked peers. If so,
+            // give them a 50% chance to be picked over all the other peers.
+            if (this.peerAddresses.Handshaked().Any())
+            {
+                int chance = this.random.Next(100);
+                if (chance <= 50)
+                {
+                    this.logger.LogTrace("(-)[RETURN_HANDSHAKED]");
+                    return this.peerAddresses.Handshaked();
```

Comment:
Is there any randomness in the HandShaked() method?

Reply:
No, a random peer is picked in `SelectPeer` from whatever set of peers is returned in `SelectPreferredPeers`.